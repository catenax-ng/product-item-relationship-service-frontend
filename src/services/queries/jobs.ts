import { useMutation } from "@tanstack/react-query";
import { useQuery } from "../../lib/react-query";
import { IRSRequestBody } from "../../types/jobs";
import { getCurrentEnvironment } from "../../utils/sessionStorageHandling";
import { selectAPI } from "../api/jobs.api";

export const useFetchJobById = ({ id }: { id: string }, refetchInterval: false | number = false) => {
  const serverEnv = getCurrentEnvironment();
  const jobAPI = selectAPI(serverEnv);
  return useQuery(["jobs", serverEnv, id], () => jobAPI.fetchJobById(id), {
    refetchInterval,
  });
};

export const useFetchJobs = (page: number, refetchInterval: false | number = false) => {
  const serverEnv = getCurrentEnvironment();
  const jobAPI = selectAPI(serverEnv);
  return useQuery(["jobs", serverEnv, page], () => jobAPI.fetchJobs(page), { refetchInterval });
};

export const useCancelJobs = () => {
  const serverEnv = getCurrentEnvironment();
  const jobAPI = selectAPI(serverEnv);
  return useMutation(jobAPI.cancelJob);
};

export const useCreateJob = () => {
  const serverEnv = getCurrentEnvironment();
  const jobAPI = selectAPI(serverEnv);
  return useMutation({
    mutationFn: (data: IRSRequestBody) => {
      return jobAPI.createJob(data);
    },
  });
};
