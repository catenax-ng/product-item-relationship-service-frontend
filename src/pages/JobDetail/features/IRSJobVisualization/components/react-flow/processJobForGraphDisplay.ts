import { Edge, Node } from "reactflow";
import { JobResponse, Relationship, Shell } from "../../../../../../types/jobs";
import { getLayoutedElements } from "./LayoutElements";

const NODE_WIDTH = 300;
const getNodeBoxHeight = (shell: Shell): number => {
  const INFO_BOX_HEIGHT = 80;
  const ASPECTS_TITLE_HEIGHT = 50;
  const ASPECTS_BUTTON_HEIGHT = 50 + 5;
  const TOTAL_BUTTON_HEIGHT = shell.submodelDescriptors.length * ASPECTS_BUTTON_HEIGHT;
  return INFO_BOX_HEIGHT + ASPECTS_TITLE_HEIGHT + TOTAL_BUTTON_HEIGHT;
};

export type GraphNodeData = Node<Shell>;
export type GraphEdgeData = Edge<Relationship>;
const getNodes = (job: JobResponse): GraphNodeData[] => {
  return job.shells.map((shell) => {
    return {
      id: shell.globalAssetId.value[0],
      data: shell,
      position: { x: 0, y: 0 },
      type: "displayNode",
      height: getNodeBoxHeight(shell),
      width: NODE_WIDTH,
    };
  });
};
const getEdges = (job: JobResponse): GraphEdgeData[] => {
  const validNodeIds = job.shells.map((x: Shell) => {
    return x.globalAssetId.value[0];
  });
  const edgeData: GraphEdgeData[] = [];
  job.relationships.forEach((rel) => {
    const from = rel.catenaXId;
    const to = rel.linkedItem.childCatenaXId;

    if (validNodeIds.includes(from) && validNodeIds.includes(to)) {
      edgeData.push({
        id: `${rel.catenaXId}-${rel.linkedItem.childCatenaXId}`,
        source: rel.catenaXId,
        target: rel.linkedItem.childCatenaXId,

        data: rel,
      });
    }
  });
  return edgeData;
};

export const processJobForGraphDisplay = (job: JobResponse) => {
  const jobNodes = getNodes(job);
  const jobEdges = getEdges(job);
  return getLayoutedElements(jobNodes, jobEdges);
};
