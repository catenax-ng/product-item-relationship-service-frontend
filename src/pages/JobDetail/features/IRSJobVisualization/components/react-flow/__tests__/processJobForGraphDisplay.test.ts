import { JobsDemoDataSuccess } from "../../../../../../../services/api/__test__/jobs.exampleData";
import { processJobForGraphDisplay } from "../processJobForGraphDisplay";

it("processes a job and sets positions for the display elements", () => {
  const job = JobsDemoDataSuccess;
  expect(processJobForGraphDisplay(job)).toMatchSnapshot();
});
