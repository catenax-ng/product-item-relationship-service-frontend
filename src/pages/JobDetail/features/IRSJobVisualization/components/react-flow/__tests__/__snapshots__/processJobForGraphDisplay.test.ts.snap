// Vitest Snapshot v1

exports[`processes a job and sets positions for the display elements 1`] = `
{
  "edges": [],
  "nodes": [
    {
      "data": {
        "description": [
          {
            "language": "en",
            "text": "The shell for a vehicle",
          },
        ],
        "globalAssetId": {
          "value": [
            "urn:uuid:a45a2246-f6e1-42da-b47d-5c3b58ed62e9",
          ],
        },
        "idShort": "future concept x",
        "identification": "882fc530-b69b-4707-95f6-5dbc5e9baaa8",
        "specificAssetIds": [
          {
            "key": "engineserialid",
            "value": "12309481209312",
          },
        ],
        "submodelDescriptors": [
          {
            "description": [
              {
                "language": "en",
                "text": "Provides base vehicle information",
              },
            ],
            "endpoints": [
              {
                "interface": "HTTP",
                "protocolInformation": {
                  "endpointAddress": "https://catena-x.net/vehicle/basedetails/",
                  "endpointProtocol": "HTTPS",
                  "endpointProtocolVersion": "1.0",
                },
              },
            ],
            "idShort": "vehicle base details",
            "identification": "4a738a24-b7d8-4989-9cd6-387772f40565",
            "semanticId": {
              "value": [
                "urn:bamm:com.catenax.vehicle:0.1.1",
              ],
            },
          },
          {
            "description": [
              {
                "language": "en",
                "text": "Provides base vehicle information",
              },
            ],
            "endpoints": [
              {
                "interface": "HTTP",
                "protocolInformation": {
                  "endpointAddress": "https://catena-x.net/vehicle/partdetails/",
                  "endpointProtocol": "HTTPS",
                  "endpointProtocolVersion": "1.0",
                },
              },
            ],
            "idShort": "vehicle part details",
            "identification": "dae4d249-6d66-4818-b576-bf52f3b9ae90",
            "semanticId": {
              "value": [
                "urn:bamm:com.catenax.vehicle:0.1.1#PartDetails",
              ],
            },
          },
        ],
      },
      "height": 240,
      "id": "urn:uuid:a45a2246-f6e1-42da-b47d-5c3b58ed62e9",
      "position": {
        "x": 0,
        "y": 0,
      },
      "sourcePosition": "bottom",
      "targetPosition": "top",
      "type": "displayNode",
      "width": 300,
    },
  ],
}
`;
